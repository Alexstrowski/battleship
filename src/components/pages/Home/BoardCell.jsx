import { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { SVG } from 'assets/svg';
import { updateSunkStatus } from 'utils/gameboard';

const BoardCell = ({
    row,
    column,
    isShip,
    isGameStarted,
    setTurnCounter,
    ship,
    ships,
    setShips,
    isRecord,
    shotsMissed,
    setShotsMissed,
}) => {
    const [selected, isSelected] = useState(
        isRecord && shotsMissed.some((ship) => ship.x === row && ship.y === column) ? true : false,
    );
    const [isSunken, setIsSunken] = useState(false);

    useEffect(() => {
        setIsSunken(ship ? ship.isSunken : false);
    }, [ship]);

    const onClickCell = () => {
        if (!isGameStarted) return;

        if (!selected) {
            setTurnCounter((prevTurn) => prevTurn - 1);

            if (isShip) {
                const id = ships.find((ship) => ship.x === row && ship.y === column).id;
                const updateShips = ships.map((ship) => {
                    return ship.x === row && ship.y === column ? { ...ship, isHit: true } : ship;
                });
                setShips(updateSunkStatus(updateShips, id));
            } else {
                setShotsMissed([...shotsMissed, { x: row, y: column }]);
            }
        }
        isSelected(true);
    };

    return (
        <div
            className="md:h-20 md:w-20 h-10 w-10 md:p-4 bg-white border-white border-2 mr-1 mb-1 rounded-md cursor-pointer"
            onClick={onClickCell}
        >
            {!isRecord ? (
                isSunken ? (
                    <div className="w-full h-full">
                        <img src={SVG[`sunken_ship_${ship.type}`]}></img>
                    </div>
                ) : selected && isShip ? (
                    <div className="w-full h-full">
                        <img src={SVG.explosion}></img>
                    </div>
                ) : selected && !isShip ? (
                    <div className="bg-gray-300 w-full h-full"></div>
                ) : (
                    ''
                )
            ) : ship && ship.isSunken ? (
                <div className="w-full h-full">
                    <img src={SVG[`sunken_ship_${ship.type}`]}></img>
                </div>
            ) : ship && ship.isHit ? (
                <div className="w-full h-full">
                    <img src={SVG.explosion}></img>
                </div>
            ) : selected && !isShip ? (
                <div className="bg-gray-300 w-full h-full"></div>
            ) : (
                ''
            )}
        </div>
    );
};

export default BoardCell;

BoardCell.propTypes = {
    row: PropTypes.number.isRequired,
    column: PropTypes.number.isRequired,
    isShip: PropTypes.bool.isRequired,
    isGameStarted: PropTypes.bool.isRequired,
    setTurnCounter: PropTypes.func,
    ship: PropTypes.object,
    ships: PropTypes.any,
    setShips: PropTypes.func,
    isRecord: PropTypes.bool.isRequired,
    shotsMissed: PropTypes.array,
    setShotsMissed: PropTypes.func,
};
