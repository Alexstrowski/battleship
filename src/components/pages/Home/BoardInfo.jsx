import PropTypes from 'prop-types';
import { Button } from 'components/ui';
import { SVG } from 'assets/svg';
const iconLevels = {
    Easy: '😀',
    Medium: '😐',
    Hard: '😡',
    Custom: '🤠',
};

const BoardInfo = ({ level, turns, isGameStarted, setShowStartPanel, gameFinishedMessage }) => {
    const restartGame = () => {
        setShowStartPanel(true);
        window.location.reload();
    };

    return (
        <div className="flex flex-col items-center rounded border-4 border-blue-600 m-10 p-10 h-full w-1/5">
            <div className="text-3xl font-bold mb-2">Difficulty: </div>
            <div className="text-3xl font-bold mb-10">
                {level.shortName} {iconLevels[level.shortName]}
            </div>
            <div className="text-3xl font-bold mb-2">Turns: </div>
            <div className="text-4xl font-bold mb-10">{turns}</div>
            <div className="flex w-full justify-around mb-2">
                <div className="flex items-center">
                    <img className="h-12 w-12 mr-1" src={SVG.sunken_ship_1}></img>
                    <div className="flex text-xl font-bold text-gray-400">x4</div>
                </div>
                <div className="flex items-center">
                    <img className="h-12 w-12 mr-1" src={SVG.sunken_ship_2}></img>
                    <div className="text-xl font-bold text-gray-400">x3</div>
                </div>
            </div>
            <div className="flex w-full justify-around mb-10">
                <div className="flex items-center">
                    <img className="h-12 w-12 mr-1" src={SVG.sunken_ship_3}></img>
                    <div className="flex text-xl font-bold text-gray-400">x2</div>
                </div>
                <div className="flex items-center">
                    <img className="h-12 w-12 mr-1" src={SVG.sunken_ship_4}></img>
                    <div className="text-xl font-bold text-gray-400">x1</div>
                </div>
            </div>
            <div className={`text-xl font-bold mb-10 text-center ${gameFinishedMessage.color}`}>
                {gameFinishedMessage.message}
            </div>

            {!isGameStarted && <Button onClick={restartGame}>Try again !</Button>}
        </div>
    );
};

export default BoardInfo;

BoardInfo.propTypes = {
    level: PropTypes.any,
    turns: PropTypes.number.isRequired,
    isGameStarted: PropTypes.bool.isRequired,
    setShowStartPanel: PropTypes.func.isRequired,
    gameFinishedMessage: PropTypes.object.isRequired,
};
