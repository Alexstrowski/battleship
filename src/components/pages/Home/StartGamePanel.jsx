import { Button, Select } from 'components/ui';
import { getLevels } from 'utils/constants/settings';
import PropTypes from 'prop-types';

const StartGamePanel = ({ initGame, level, chooseLevel, setShowStartPanel }) => {
    const handleStartGame = () => {
        initGame(true);
        setShowStartPanel(false);
    };

    const handleChangeLevel = (level) => {
        chooseLevel(level);
    };
    return (
        <div className="flex flex-col items-center rounded border-4 border-blue-600 m-10 p-10 h-full w-1/5">
            <div className="text-xl font-bold mb-2">Select your difficulty</div>
            <div className="mb-4">
                <Select dataList={getLevels()} selected={level} onChange={handleChangeLevel} />
            </div>
            <Button onClick={handleStartGame}>Start!</Button>
        </div>
    );
};

export default StartGamePanel;

StartGamePanel.propTypes = {
    initGame: PropTypes.func.isRequired,
    level: PropTypes.object.isRequired,
    chooseLevel: PropTypes.func.isRequired,
    setShowStartPanel: PropTypes.func.isRequired,
};
