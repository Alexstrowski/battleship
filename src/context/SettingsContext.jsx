import { createContext } from 'react';
import useLocalStorage from 'hooks/useLocalStorage';
import PropTypes from 'prop-types';
import { getLevels, CUSTOM_TURNS_BY_DEFAULT } from 'utils/constants/settings';

const SettingsContext = createContext();
const levels = getLevels();

const SettingsProvider = ({ children }) => {
    const [customTurns, setCustomTurns] = useLocalStorage('customTurns', CUSTOM_TURNS_BY_DEFAULT);
    const [defaultLevel, setDefaultLevel] = useLocalStorage('level', levels[0]);

    const updateCustomTurns = (turns) => {
        setCustomTurns(turns);
    };

    const updateLevel = (level) => {
        setDefaultLevel(level);
    };

    const data = { customTurns, defaultLevel, updateCustomTurns, updateLevel };

    return <SettingsContext.Provider value={data}>{children}</SettingsContext.Provider>;
};

export { SettingsProvider, SettingsContext };

SettingsProvider.propTypes = {
    children: PropTypes.node.isRequired,
};
