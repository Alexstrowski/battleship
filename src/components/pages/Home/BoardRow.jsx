import BoardCell from './BoardCell';
import PropTypes from 'prop-types';

const BoardRow = (props) => {
    const { row, ships } = props;
    return (
        <div className="flex">
            {[...Array(10).keys()].map((column) => {
                const isShip = ships.some((shipPosition) => shipPosition.x === row && shipPosition.y === column);
                const ship = ships.find((ship) => ship.x === row && ship.y === column);
                return <BoardCell key={column} isShip={isShip} column={column} ship={ship} {...props} />;
            })}
        </div>
    );
};

export default BoardRow;

BoardRow.propTypes = {
    row: PropTypes.number.isRequired,
    ships: PropTypes.array.isRequired,
};
