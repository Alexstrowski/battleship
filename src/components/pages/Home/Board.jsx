import BoardRow from './BoardRow';
import { getAllShips } from 'utils/gameboard';

const arrayy = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
const ships = getAllShips();
const Board = () => {
    return (
        <div className="p-5 bg-blue-400 w-max mt-10">
            {arrayy.map((i) => (
                <BoardRow key={i} i={i} ships={ships} />
            ))}
        </div>
    );
};

export default Board;
