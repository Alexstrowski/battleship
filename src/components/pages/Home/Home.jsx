import { useState, useEffect } from 'react';
import Board from './Board';
import StartGamePanel from './StartGamePanel';
import BoardInfo from './BoardInfo';
import { LEVELS } from 'utils/constants/settings';
import useLocalStorage from 'hooks/useLocalStorage';
import { getAllShips } from 'utils/gameboard';

const messages = {
    win: { message: 'You destroyed all the ships, you win.', color: 'text-green-500' },
    lose: { message: 'You did not destroy all the ships, you lose.', color: 'text-red-500' },
    none: { message: '', color: '' },
};
const randomShips = getAllShips();

const Home = () => {
    const [ships, setShips] = useState(randomShips);
    const [isGameStarted, setIsGameStarted] = useState(false);
    const [level, setLevel] = useState(LEVELS[0]);
    const [turnCounter, setTurnCounter] = useState(LEVELS[0].turns);
    const [localStorageBoardHistory, setLocalStorageBoardHistory] = useLocalStorage('boardHistory', []);
    const [showStartPanel, setShowStartPanel] = useState(true);
    const [gameFinishedMessage, setGameFinishedMessage] = useState(messages.none);

    console.log(ships);

    const validateEndGame = () => {
        const isAllShipsDestroyed = ships.every((ship) => ship.isSunken === true);
        if (isAllShipsDestroyed) {
            setGameFinishedMessage(messages.win);
            setIsGameStarted(false);
            setLocalStorageBoardHistory([...localStorageBoardHistory, ships]);
            return;
        }

        if (turnCounter === 0) {
            setGameFinishedMessage(messages.lose);
            setIsGameStarted(false);
            setLocalStorageBoardHistory([...localStorageBoardHistory, ships]);
            return;
        }
    };

    useEffect(() => {
        setTurnCounter(level.turns);
    }, [level]);

    useEffect(() => {
        validateEndGame();
    }, [turnCounter]);

    return (
        <div className="container mx-auto">
            <div className="flex justify-center">
                <Board
                    isGameStarted={isGameStarted}
                    setTurnCounter={setTurnCounter}
                    ships={ships}
                    setShips={setShips}
                    isRecord={false}
                />

                {showStartPanel ? (
                    <StartGamePanel
                        initGame={setIsGameStarted}
                        level={level}
                        chooseLevel={setLevel}
                        setShowStartPanel={setShowStartPanel}
                    />
                ) : (
                    <BoardInfo
                        level={level}
                        turns={turnCounter}
                        isGameStarted={isGameStarted}
                        setShips={setShips}
                        setShowStartPanel={setShowStartPanel}
                        gameFinishedMessage={gameFinishedMessage}
                    />
                )}
            </div>
        </div>
    );
};

export default Home;
