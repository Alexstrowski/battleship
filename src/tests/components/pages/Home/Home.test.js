import { shallow, mount } from 'enzyme';
import Home from 'components/pages/Home';
import { SettingsProvider } from 'context/SettingsContext';

describe('Tests in <Home />', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(
            <SettingsProvider>
                <Home />
            </SettingsProvider>,
        );
    });

    test('Must render <Home /> correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('The game should start correctly', () => {
        let wrapperMount = mount(
            <SettingsProvider>
                <Home />
            </SettingsProvider>,
        );
        const buttonComponent = wrapperMount.find('#start-button').at(0);
        buttonComponent.find('button').at(0).simulate('click');
        const difficultyText = wrapperMount.find('p').at(0).text().trim();
        const turnsText = wrapperMount.find('p').at(1).text().trim();
        expect(difficultyText).toBe('Difficulty:');
        expect(turnsText).toBe('Turns:');
    });

    test('The table must have 100 cells', () => {
        let wrapperMount = mount(
            <SettingsProvider>
                <Home />
            </SettingsProvider>,
        );
        const boardCell = wrapperMount.find('#board-cell');
        expect(boardCell.length).toBe(100);
    });
});
