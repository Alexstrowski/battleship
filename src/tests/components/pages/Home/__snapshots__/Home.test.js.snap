// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <Home /> Must render <Home /> correctly 1`] = `
<ContextProvider
  value={
    Object {
      "customTurns": 45,
      "defaultLevel": Object {
        "id": 1,
        "name": "Easy - Infinity turns",
        "shortName": "Easy",
        "turns": 200,
      },
      "updateCustomTurns": [Function],
      "updateLevel": [Function],
    }
  }
>
  <Home />
</ContextProvider>
`;
